protocol/0053-PERP-product_builtin_perpetual_future.md

```diff
@@ -28,23 +61,84 @@ cash_settled_perpetual_future.value(quote) {
 
 ```javascript
-cash_settled_perpetual_future.settlement_data(event) {
-	cashflow = cash_settled_perpetual_future.value(event.data) - cash_settled_perpetual_future.value(market.mark_price))
-	settle(cash_settled_perpetual_future.settlement_asset, cashflow)
-	setMarkPrice(event.data)
+cash_settled_perpetual_future.settlement_cue(event) {
+	return vegaprotocol.builtin.timestamp
+}
+```
+
+### 4.2. Periodic settlement data point received
+
+Once the [settlement cue](#41-periodic-settlement-cue) time is received the specified settlement data oracle gets monitored for incoming data. It must be possible to filter the data by the time it was received (see [0048-DSRI-data_source_internal](./0048-DSRI-data_source_internal.md#13-vega-time-changed)) at and to use the time provided by the settlement cue as a variable in such filter. If the periodic settlement data received satisfies all the filters that have been specified for it then that data point (`y`) along with the current `mark_price` (`x`) for the market and the current `vegaprotocol.builtin.timestamp` (`t`) gets stored as the funding rate data point.
+
+### 4.3. Mark to market settlement
+
+Every time a [mark to market settlement](./0003-MTMK-mark_to_market_settlement.md) is carried out the value of the last periodic settlement data point received along with the price used for MTM settlement and the current `vegaprotocol.builtin.timestamp` gets stored as the funding rate data point. If no periodic settlement data has been received yet then the funding rate data point should not be created.
+
+### 4.4. Periodic settlement
+
+When the `settlement_schedule` event is received the latest funding rate data point gets repeated with the timestamp set to the current value of `vegaprotocol.builtin.timestamp`.
+
+The next step is to calculate the periodic settlement funding rate. If there are no periodic settlement data points then the periodic settlement is skipped. Otherwise, consider all the periodic settlement data points and calculate the time-weighted average price difference as:
+
+```go
+sd := 0
+st := 0
+for i := 0; i < len(data_points) - 1; i++ {
+    t := data_points[i+1].t-data_points[i].t
+    d := data_points[i].x - data_points[i].y
+    sd += d * t
+    st += t
 }
+funding_rate = sd / st
 ```
 
+All the funding rate data points except for the last one (it should get carried over as the first data point for the next period) can then be deleted.
+
+Last step is to calculate each party's cash flows as $-\text{open volume} * \text{funding rate}$ where cashflows are first collected from parties that are making the payment (negative value of the cashflow, i.e. longs when the funding rate is positive) and distributed to those receiving it. Any shortfall should be made-up from the insurance pool and if that's not possible loss socialisation should be applied (exactly as per mark-to-market settlement methodology).
```

Comment:
Do we want to use the insurance pool for funding rate cashflow shortfalls? Entire amount or a fraction to discourage oracle manipulation?

Reply:
I'd say we just make the entire insurance pool available. The data sourcing (in the future) should deal with making the data valid; on the level of the product we shouldn't care.